interfaces/greengrass_support: update accesses for GGC 1.8

anonymouse64 opened the pull request:
Testing this interface with the snap is quite complicated, please reach out to me over IRC/email if you would like information on how to test the interface directly with the snap.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -169,12 +332,16 @@ const greengrassSupportConnectedPlugSeccomp = `
 setgroups32
 
+# for joining another namespace
+setns
```
Which ones? From network-control you get:
```
setns - CLONE_NEWNET
```

Please enumerate the namespaces greengrass is using (you can see these via strace) or if it is using all of them, mention that in the comment.

anonymouse64 replied:
The only namespace I see with strace is just `CLONE_NEWNS`. I'll update the rule here to just add that, but how does the seccomp compiler handle multiple syscall lines with different arguments? Do they get combined in the rule that runs or do they conflict with each other?

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+# specific rule for cpuset files
+owner /old_rootfs/sys/fs/cgroup/cpuset/{,system.slice/,user.slice/,user.slice/system.slice/}cpuset.{cpus,mems} rw,
+# TODO: ideally we could include "net_cls,net_prio" inside the one above this that handles
+# blkio, cpuset, memory, etc. but it's not currently clear how to escape a comma inside the pattern...
```
Thanks for the tip, and also done; comment removed.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
-mount fstype=proc proc -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/proc/,
+# cgroup accesses
+# greengrassd extensively uses cgroups and needs to read what cgroups are available
+/proc/cgroups r,
```
Done

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+owner /old_rootfs/sys/fs/cgroup/pids/user.slice/user-1000.slice/{,user@1000.service/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/pids/user.slice/user-1000.slice/{,user@1000.service/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+owner /old_rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
```
These accesses only appear when running the greengrass snap on Ubuntu classic (i.e. specifically Ubuntu 18.04 classic), and also as a command rather than a service. 

I originally only had the `greengrassd` daemon run as an app in the snap, because to run it needs to have certificates installed by the user at runtime. If the certificates weren't available when `greengrassd` runs, it would fail, which means that if `greengrassd` was a proper service in the snap, the snap couldn't be installed as the service would fail to start during install and the install would be aborted. 

I debated removing these now that the necessary functionality to disable services from the install hook landed in snapd 2.36 (see PR https://github.com/snapcore/snapd/pull/5979), and we have the `greengrassd` daemon as a proper service in the snap, but I figured if some day the greengrass snap is adjusted to be a command the user invokes manually again then we may want these accesses.

All that being said, yes Greengrass does appear to just try to read all cgroups (and it does fail if any of the reads are denied), so if you'd prefer we could just use a simpler rule instead.

jdstrand replied:
I suggest having policy that works when running greengrassd as a service since that is how it will be delivered. I suspect it will clean up a few things in the policy. The older snap had instructions for setting up the certificates. You could disable the services like you suggest or simply check if the certificates are present and if not log that the service wasn't started because the certs weren't present. Up to you on how to implement the snap, but this seems like a clean way forward policy-wise.

anonymouse64 replied:
I have adjusted this all to only include write accesses to cgroup paths when running as a systemd service.

anonymouse64 replied:
(also I added a rule for read access anywhere and only have write specific paths for the ones it seems to use)

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+owner /old_rootfs/sys/fs/cgroup/cpu,cpuacct/{,user.slice/,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/cpu,cpuacct/{,user.slice/,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+owner /old_rootfs/sys/fs/cgroup/pids/user.slice/user-1000.slice/{,user@1000.service/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/pids/user.slice/user-1000.slice/{,user@1000.service/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
```
See my previous comment about running `greengrassd` as an app rather than a service. I don't think it's for the ggc_user, I just think it's for all cgroup accesses it can possibly read.

anonymouse64 replied:
I removed these accesses

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+# blkio, cpuset, memory, etc. but it's not currently clear how to escape a comma inside the pattern...
+owner /old_rootfs/sys/fs/cgroup/net_cls,net_prio/{,user.slice/,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/net_cls,net_prio/{,user.slice/,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+# same comment as for "net_cls,net_prio" above but for "cpu,cpuacct" ...
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+/proc/sys/fs/protected_hardlinks r,
+/proc/sys/fs/protected_symlinks r,
+
+
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+mount options=(rw, bind) /var/lib/extrausers/passwd ->
+/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/etc/passwd,
+mount options=(rw, bind) /var/lib/extrausers/group ->
+/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/etc/group,
```
Well AFAICT greengrass only uses the `/etc/{passwd,group}` file for looking up which uid/gid corresponds to `ggc_user` or `ggc_group`, but in my testing this mount works properly as far as running lambda functions inside the snap goes. If you like, I could instead do as you propose and merge the two files, and mount that merged file inside the mount namespace we prepare for `greengrassd`.

jdstrand replied:
I think merging them is going to be more robust in the long run. As a side benefit, we can remove these mount rules.

anonymouse64 replied:
I implemented as suggested and just combine the two files manually inside $SNAP_DATA/{group,passwd} and mount that inside the rootfs and this works well.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+/dev/loop-control rw,
+/dev/loop[0-9]* rw,
+/sys/devices/virtual/block/loop[0-9]*/ r,
+/sys/devices/virtual/block/loop[0-9]*/** r,
```
Do you mean a comment in the Description of the interface, or just right here above the specific accesses?

jdstrand replied:
Here, but I also asked you to update the description in another comment I gave.

anonymouse64 replied:
Done

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+/bin/{,u}mount ixr,
+/sbin/pivot_root ixr,
+
+# allow pivot_root'ing into the rootfs prepared for the greengrass daemon
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+# for determining users who performed mounts
+/run/mount/utab rw,
+
+# mounts for classic systems for ggc_group and ggc_user support
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+/sys/devices/virtual/block/loop[0-9]*/ r,
+/sys/devices/virtual/block/loop[0-9]*/** r,
+
+# mounts for setting up the mount namespace that greengrassd runs inside
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +243,23 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
+umount /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/**,
+umount /ggc-writable/packages/1.6.0/rootfs_nosys/,
+umount /ggc-writable/packages/1.6.0/rootfs_sys/,
+umount /ggc-writable/packages/1.6.0/rootfs_sys/sys/,
```
Yes this should have been `*`. Done

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +243,23 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
-/state/ r,
-/state/sqlite* rwk,
+# for the greengrassd pid file
+owner /{var/,}run/greengrassd.pid rw,
```
The pid file is created by greengrassd which runs inside of a mount namespace we setup for it and pivot_root into, so this access is also in the post-pivot_root area. Shall I add a comment to reflect that here? Or perhaps more generally speaking I should move all accesses that are needed post-pivot_root to a specific section that is labeled clearly as being for post-pivot_root.

But to answer your question, no this can't be solved by layouts.

jdstrand replied:
I like the idea of moving the things that we know for sure are post-pivot_root somewhere-- it will be easier to audit and you can have a single comment talking about them.

anonymouse64 replied:
I have reorganized all the accesses to be at the end of the profile

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+# ideally we would set some sort of environment variable to force sqlite to change it's temp dir as described on
+# https://bugs.launchpad.net/ubuntu-ui-toolkit/+bug/1197049
+# it could also be possible that sqlite is pulled in from one of the stage-packages in the snap though...
+owner /var/tmp/etilqs_[0-9a-zA-Z]* rw,
```
This is not in the global namespace, it's also in the post-pivot_root area.

anonymouse64 replied:
@jdstrand is this access okay since it's post pivot_root or do we need to work on reworking the snap to avoid this? This access seems to be baked into greengrassd itself and would be difficult to remove at this point.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+
+# for certificates files - note that the pivot_root into $SNAP_DATA/rootfs means that these accesses are actually
+# meant for $SNAP_DATA/rootfs/greengrass, but can't use the actual path due to pivot_root
+owner /greengrass/ mrix,
```
Done.

anonymouse64 replied:
As mentioned above all of these accesses have been removed

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+owner /greengrass/ mrix,
+/greengrass/** mrix,
+/greengrass/certs/ rw,
+/greengrass/certs/{,**} rw,
```
Done.

anonymouse64 replied:
As mentioned above all of these accesses have been removed

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+
+# on UC16 the ggc-writable paths resolve to this in apparmor
+/system-data/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/ggc-writable/ rw,
+/system-data/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/ggc-writable/{,**} rw,
```
This is the path that AppArmor shows as being denied when running on Ubuntu Core 16. On Ubuntu classic installations, such as Ubuntu 18.04, this access doesn't appear for some reason and instead shows up as `/var/snap/...` I suppose I could change the `/var/snap` rules to all be `{,/system-data}/var/snap/...` but for some reason this is the only path that gets resolved to `/system-data`.

jdstrand replied:
Do we expect this snap to be usable on classic distro? If so, I suggest grouping all the classic-specific accesses together, and conditionally add those using `if release.OnClassic {`. dbus.go does something like this now (along with others). You might want to also look at x11.go for inspiration. Be sure to add unit tests that the rules are present when mocking classic and aren't when not.

anonymouse64 replied:
Yes Amazon expects the snap to be usable on classic as well. That's a good idea about conditionally adding the necessary classic accesses, thanks for the pointer.

anonymouse64 replied:
This access still happens, so should I adjust this bit to only be added with logic like `if !release.OnClassic { ...`?

anonymouse64 replied:
I now add the snippet conditionally to the apparmor snippet and have a test in place for this.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+# for child container lambdas - note these are also actually not in the root folder, but rather mounted
+# from $SNAP_DATA/rootfs/greengrass/* to / for the individual containers
+/certs/ r,
+/certs/** r,
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+# the child containers need to use a file lock here
+owner /state/secretsmanager/secrets.db krw,
+owner /state/secretsmanager/secrets.db-journal rw,
+owner /state/shadow/ krw,
```
Deleted.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+owner /state/shadow/ krw,
+owner /state/shadow/{,**} krw,
+# more specific accesses for writing
+owner /state/server/ rw,
```
Deleted.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+
+# the java runtime attempts to access this
+/etc/ssl/certs/java/ r,
+
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -169,12 +332,16 @@ const greengrassSupportConnectedPlugSeccomp = `
 
 # allow use of ggc_user and ggc_group
 # FIXME: seccomp arg filter by this uid/gid when supported by snap-confine
+lchown
```
Done.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -188,18 +355,29 @@ sethostname
+# device creation is necessary for mounting a squashfs file using loopback devices
+mknod
+mknodat
+
```
Done.

pedronis commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+# greengrassd extensively uses cgroups and needs to read what cgroups are available
+/proc/cgroups r,
+owner @{PROC}/[0-9]*/cgroup r,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,user.slice/,system.slice/}system.slice/ rw,
```
I see two greengrass named snaps (greengrass, aws-greengrass) but no snap, no even those have a snap-declaration asking for greengrass-support atm

anonymouse64 replied:
As there are no current snaps using `greengrass-support` I think we are okay to break this interface as much as needed to get the initial version out. After the initial version obviously we will have to worry about breaking previous revisions however.

zyga commented:
There's a lot of unit test failures. Please fix locally and push.

anonymouse64 commented:
Just to update on this, I have spoken with AWS and while they won't be implementing support for their daemon to set child profiles for the containers in the immediate timeframe, they are evaluating it and I'm helping them with it. Also I ran into an issue with my changes here when using the greengrass snap with other interfaces such as `gpio` connected that do trigger udev rules (and thus also the device cgroup being enabled). While we don't need to have these other interfaces auto-connected, AWS wants to be able to plug these device specific interfaces available if a customer chooses (either brand store or otherwise).

I discussed with @jdstrand a while ago over IRC about implementing a flag for interfaces to declare that they control their own device cgroup and shouldn't have udev tagging rules generated (and thus also not have the device cgroup enforced). I will update this PR with a patch I have that does this (I did get it to work, just need tests, etc.). This means that I can remove the trick for including the `network-control` AppArmor and Seccomp into the `greengrass-support` interface and just have the snap plug `network-control` normally. Not sure if I will finish before the field/sales sprint is over though, so I just thought I would update here the status first.

Please let me know if I should start a design discussion on the forum related to this spec which declares an interface as controlling device cgroups and I will do so.

pedronis commented:
@anonymouse64 yes a spec  on the forum that can be vetted there and given context with @jdstrand and leave a trail seems appropriate for that

anonymouse64 commented:
@pedronis forum post here: https://forum.snapcraft.io/t/allowing-interfaces-to-disable-device-cgroup-udev/9630 Thanks! Should I close this and just re-open after discussion there is complete?

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -188,18 +355,29 @@ sethostname
+
+# device creation is necessary for mounting a squashfs file using loopback devices
+mknod
+mknodat
```
@jdstrand I actually looked into this more and the access needed is not `S_IFBLK`, it's actually `S_IFCHR`. The syscall looks like this:

```
mknodat(AT_FDCWD, "/var/snap/greengrass/x1/ggc-writable/packages/1.7.0/rootfs/upper/snap/greengrass/x1", S_IFCHR|000, makedev(1, 3)) = -1 EPERM (Operation not permitted)
```
(found with `sudo snap run --strace="-f -emknodat"`

`S_IFBLK` doesn't appear to be used at all during a normal deployment AFAICT. I see from the snap-confine wiki that OR'ing multiple bits isn't yet implemented in the seccomp compiler, so if we do encounter a `S_IFBLK` (or other node type) in the wild, wouldn't we need to change this to `mknodat` (or implement OR'ing multiple bits together in the compiler)?

anonymouse64 replied:
Also I updated the descriptions as requested

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+owner /old_rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/session-[0-9]*.scope/system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/session-[0-9]*.scope/system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
+owner /old_rootfs/sys/fs/cgroup/{devices,memory,pids,blkio,systemd}/{,user.slice/,user.slice/system.slice/,system.slice/}snap.greengrass.greengrass{,d.service}/system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/{devices,memory,pids,blkio,systemd}/{,user.slice/,user.slice/system.slice/,system.slice/}snap.greengrass.greengrass{,d.service}/system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a- ...
```
I added a comment about it and also about the eventual work to get stacking cgroups to work

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+
+
+# for determining users who performed mounts
+/run/mount/utab rw,
```
It still works with the deny, so I added the deny.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+
+# greengrass 1.7 mounts
+mount options=(rw, rbind) /ggc-writable/packages/*/rootfs/merged/ -> /ggc-writable/packages/*/rootfs/merged/,
+mount options=(rw, bind) /greengrass/ggc/packages/*/lambda/GreengrassSystemComponents/ -> /ggc-writable/packages/*/rootfs/merged/lambda/,
```
I actually rearchitected the snap to not use any of these root filesystem accesses and now inside the middle-layer mount namespace everything is mounted as if it was running in the normal snap mount namespace with files at $SNAP, and the writable directory in $SNAP_DATA so none of these `ggc*` accesses are necessary anymore.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
-mount options=(rw, bind) /dev/random -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/dev/random,
-mount options=(rw, bind) /dev/tty -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/dev/tty,
-mount options=(rw, bind) /dev/urandom -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/dev/urandom,
-mount options=(rw, bind) /dev/zero -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/dev/zero,
```
See above response

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+
+# for using the ggc-writable directory
+/ggc-writable/ rw,
+/ggc-writable/{,**} rw,
```
As mentioned above all of these accesses have been removed

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+/system-data/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/ggc-writable/{,**} rw,
+
+# for child container lambdas - note these are also actually not in the root folder, but rather mounted
+# from $SNAP_DATA/rootfs/greengrass/* to / for the individual containers
```
I have discussed this with AWS and they are evaluating the option and I have added a comment to reflect this in the profile.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+# the python runtime tries to access /etc/debian_version, presumably to identify what system it's running on
+# note there may be other accesses that the containers try to run...
+/etc/ r,
+/etc/debian_version r,
```
I think so, I added it

anonymouse64 replied:
Yeah I for some reason didn't get it in my commit, but I added it back now.

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,6 +269,60 @@ umount /var/snap/@{SNAP_NAME}/**,
+/etc/debian_version r,
+
+# the java runtime attempts to access this
+/etc/ssl/certs/java/ r,
```
Did you mean `/etc/ssl/certs/java/{,**} r,` ? Also yes I added the `abstractions/ssl_certs`

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -188,18 +355,29 @@ sethostname
+	// this is needed because plugging network-control normally by the greengrass snap
+	// will trigger a device cgroup to be enforced for greengrass which
+	// causes greengrass to fail, as greengrass needs to manage cgroups for it's
+	// containers itself
```
This is no longer necessary, as we will now have the greengrass-support interface disable all udev tags and thus disable the device cgroup from ever being enabled. This will be implemented in a followup PR however.

anonymouse64 commented:
Since this was getting rather large, I rebased my PR and incorporated all of the feedback with the exception of the outstanding questions I had. The tests also pass locally, so hopefully travis agrees. I will file a followup PR for the device cgroup disabling feature.

anonymouse64 commented:
Looks like the tests timed out:

```
2019-02-12 17:59:39 Preparing google:fedora-29-64:tests/main/interfaces-snapd-control-with-manage...


The job exceeded the maximum time limit for jobs, and has been terminated.
```

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +75,166 @@ capability dac_read_search,
+/proc/cgroups r,
+owner @{PROC}/[0-9]*/cgroup r,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,user.slice/,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,user.slice/,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9 ...
```
I added this in the comment

anonymouse64 commented:
The followup PR implementing a spec to disable udev tagging (and thus the device cgroup) is here: https://github.com/snapcore/snapd/pull/6494

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +94,134 @@ capability dac_read_search,
+owner @{PROC}/[0-9]*/cgroup r,
+owner /old_rootfs/sys/fs/cgroup/{,**} r,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f] ...
```
`/{,system.slice/}system.slice` evaluates to /system.slice or system.slice/system.slice-- is that intended? Is it this way because of the snap.greengrass alternative? If so, I wonder if you can have a couple snap.greengrass rules and a couple non-snap.greengrass rules instead of both on the same line. (I like to use the {a,b} alternates when it makes things still easy to read/audit.

You are also allowing writes to the system blkio,..., was that intentional?

The thoughts behind these questions apply to the next set of cgroup rules too. Please consider if they could be written more clearly while being somewhat concise.

anonymouse64 replied:
So this is basically the full set of accesses I see for first Ubuntu 18.04 classic:

```
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.cpus
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.mems
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
```

and on Ubuntu 16.04 classic:

```
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.cpus
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.mems
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
```

and on Ubuntu Core 16:
```
/old_rootfs/sys/fs/cgroup/blkio/system.slice/system.slice/
/old_rootfs/sys/fs/cgroup/blkio/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/system.slice/
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/memory/system.slice/system.slice/
/old_rootfs/sys/fs/cgroup/memory/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/pids/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
```

I do not know why some of the cgroups are for `system.slice/snap.greengrass.greengrassd.service` on classic 18.04 and some are for `system.slice/system.slice` on 16.04 versions, the only thing I can think of is different kernel and thus different cgroup versions? Or perhaps different systemd versions?

jdstrand replied:
Weird. Those uuid rules are fairly terrible. AppArmor or snapd could add a @{uuid} variable to make these easier to read. That isn't for this PR.

These rules are fine.

anonymouse64 commented:
The test failures still seem to be unrelated:
```
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 linux-generic-lts-xenial : Depends: linux-image-generic-lts-xenial (= 4.4.0.143.124) but 4.4.0.141.121 is to be installed
E: Unable to correct problems, you have held broken packages.
quiet: end of output.
```

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +94,134 @@ capability dac_read_search,
+/proc/sys/fs/protected_hardlinks r,
+/proc/sys/fs/protected_symlinks r,
+
+# for determining users who performed mounts
```
The comment isn't quite right with the deny. Maybe say something about how it isn't needed.

anonymouse64 replied:
Yes I forgot to also update the comment when I turned it into a `deny`.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +94,134 @@ capability dac_read_search,
+
+# greengrassd requires protected hardlinks and symlinks 
+/proc/sys/fs/protected_hardlinks r,
+/proc/sys/fs/protected_symlinks r,
```
Please adjust to use @{PROC}. Also this only allows reading if they are set, not updating the files to do what it apparently requires (from the comment).

anonymouse64 replied:
Greengrass won't actually turn these things on itself, it will only check to make sure they are enabled. It does print a nice error message about them if it's off, and I'd rather not add rules to enable these things with sysctl.
I will update the comment to make it clear why it only needs read-access.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +94,134 @@ capability dac_read_search,
+# mounts for setting up child container rootfs
+mount options=(rw, rprivate) -> /,
+mount options=(ro, remount, rbind) -> /,
+mount fstype=overlay no_source -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/ggc-writable/packages/*/rootfs/merged/,
```
no_source? Should this be omitted (ie, `mount fstype=overlay -> /var/...`)?

anonymouse64 replied:
It can be omitted, I put `no_source` there as that's what showed up in the strace I did, but the source is there explicitly as `no_source` or if it's just empty.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -86,38 +94,134 @@ capability dac_read_search,
+# we have to manually do this otherwise the go DNS resolver fails to work, because it isn't configured to 
+# use the system DNS server and attempts to do DNS resolution itself, manually inspecting /etc/resolv.conf
+mount options=(rw, bind) /run/systemd/resolve/stub-resolv.conf -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/etc/resolv.conf,
+mount options=(rw, bind) /run/resolvconf/resolv.conf -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/rootfs/etc/resolv.conf,
```
Why rw?

anonymouse64 replied:
No good reason other than the default for `mount --bind` is rw. I tested with `mount -o ro --bind ...` and with `options=(ro, bind)` in the AppArmor and there's an additional mount that's run to remount it after it's initially setup, but this can be made read-only if you like.

jdstrand replied:
I think I'd prefer it to be read-only. This is getting into overriding the host and while greengrass has enough to do all kinds of things on the host, putting some guardrails in place seems like a good idea.

anonymouse64 replied:
This is read-only now.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +230,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
+capability mknod,
+
+# for the greengrassd pid file
+owner /{var/,}run/greengrassd.pid rw,
```
This isn't listed as post pivot root. Could you use adjust the location or use a layout? (I suspect a layout might be undesirable with all the other mounts that are happening). This isn't a blocker, but wondering if it could be in SNAP_COMMON or something.

anonymouse64 replied:
Layouts won't work for this actually for a different reason because `/run` is explicitly disallowed by layouts, and on some systems (mine included, perhaps all but I can't say for certain) `/var/run` is a symlink to `/run` and as such can't be mounted on top of.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +230,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
+# the child containers need to use a file lock here
+owner /state/secretsmanager/secrets.db krw,
+owner /state/secretsmanager/secrets.db-journal rw,
+owner /state/shadow/ krw,
```
Do you really need 'k' for this directory?

anonymouse64 replied:
After looking more closely it doesn't need to lock the directory here, but the files underneath yes.

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +230,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
+# but those version numbers could change so we add a "*" on the end of the folders to be safe for
+# future potential upgrades
+/runtime/{python*,executable*,nodejs*,java*}/ r,
+/runtime/{python*,executable*,nodejs*,java*}/** r,
```
Does it make sense to limit the runtimes? Eg, could use `/runtime/** r,`, that said, maybe this is desirable to only support blessed runtimes...

anonymouse64 replied:
I think for now it is better to only support certain runtimes as this, but that being said this is just a read-rule so it is probably also fine to allow all eventual `/runtime` subdirs. FWIW they did add a new runtime type (executable) for 1.6 or 1.7, so it's possible they will add more in the future

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +230,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
+# ideally we would set some sort of environment variable to force sqlite to change it's temp dir as described on
+# https://bugs.launchpad.net/ubuntu-ui-toolkit/+bug/1197049
+# it could also be possible that sqlite is pulled in from one of the stage-packages in the snap though...
+owner /var/tmp/etilqs_[0-9a-zA-Z]* rw,
```
Reading https://bugs.launchpad.net/ubuntu-ui-toolkit/+bug/1197049, I think if you set TMPDIR=/tmp it should work.

anonymouse64 replied:
Actually in the new 1.7 version of greengrass we will release with, it isn't necessary at all anymore

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -160,21 +291,38 @@ umount /var/snap/@{SNAP_NAME}/**,
+/etc/debian_version r,
+
+# the java runtime attempts to access this
+/etc/ssl/certs/java/{,*} r,
```
It is weird this isn't in the abstraction... :\

anonymouse64 replied:
Indeed it is odd. Perhaps a bug to file with apparmor?

jdstrand replied:
Yeah, https://bugs.launchpad.net/apparmor/+bug/1816372

anonymouse64 replied:
I added a comment about the bug to the access here.

anonymouse64 commented:
I think I addressed all the comments except the cgroup write accesses

jdstrand commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +241,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
-/bin/umount ixr,
+# for the greengrassd pid file
+# note we can't use layouts for this because /var/run is a symlink to /run
+# and /run is explictly disallowed for use by layouts
```
This is causing a test failure:
```
interfaces/builtin/greengrass_support.go:258:14: "explictly" is a misspelling of "explicitly"
```

anonymouse64 replied:
Fixed

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -35,10 +41,23 @@ const greengrassSupportBaseDeclarationSlots = `
+# /system-data/var/snap/greengrass/x1/ggc-writable/packages/1.7.0/var/worker/overlays/$UUID/upper/
+/system-data/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/ggc-writable/ rw,
+/system-data/var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/*/ggc-writable/{,**} rw,
+
```
Done, removed

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +241,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
-/bin/umount ixr,
+# for the greengrassd pid file
+# note we can't use layouts for this because /var/run is a symlink to /run
+# and /run is explictly disallowed for use by layouts
```
Done

anonymouse64 commented on interfaces/builtin/greengrass_support.go:
```diff
@@ -126,31 +241,58 @@ mount options=(rw, bind) /proc/irq/ -> /proc/irq/,
+# currently the runtimes are "python2.7", "nodejs6.10", "java8" and "executable",
+# but those version numbers could change so we add a "*" on the end of the folders to be safe for
+# future potential upgrades
+/runtime/{python*,executable*,nodejs*,java*}/ r,
```
I would rather leave it as is, simply because this is technically a leak if the host filesystem for some odd reason has things inside `/runtime` I think it's better to only support specific runtimes.

jdstrand commented:
The commit with the 1.8 changes is fine. @anonymouse64 talked to us about the ix and we are investigating it, but it's fine as is for now.
